### PR TITLE
JIT backward closure comments

### PR DESCRIPTION
**Pull request by ezyang**
Pretty substantative so I'm going to PR it. More to come.

### CODE REVIEW

File: torch/csrc/autograd/functions/jit_closure.cpp [hunk 1]
```diff
@@ -226,8 +232,57 @@ struct FusionGroupFunction : public Function {
...
+//      %33.0, %33.1 = Eval(%32, %23.1)
+//      %44.0, %44.1, %44.2, %44.3 = Eval(%33.0, %31, %10.2)
+//      %78.0, %78.1 = Eval(%44.1, %3.1)
+//      return (%6.0, %10.1, %23.0, %44.0, %44.2, %78.0);
+//    }
+//
+// Then:
+//
+//    graph->stage() = 2
+//    stage_begins = [%3, %10, %33]
```
Comment (apaszke) [inline torch/csrc/autograd/functions/jit_closure.cpp]: `stage_begins` has `graph->stage + 2` entries - the last one is an output node. There's a comment below which explains why is it this way

File: torch/csrc/autograd/functions/jit_closure.cpp [hunk 2]
```diff
@@ -226,8 +232,57 @@ struct FusionGroupFunction : public Function {
...
+//      [%6.1, %3.1],
+//      [%23.1, %10.2],
+//      [] # Always empty!
+//    ]
 struct CrossStageStateDesc {
   CrossStageStateDesc(Graph* graph)
+    // Suppose that you have a Graph which recorded traces for forwards and
+    // backwards.  Then graph->stage() == 1 (stage advancement does not occur
+    // until advanceStage()).  Then we need to record inputs from stage 0 and
+    // stage 1.
```
Comment (apaszke) [inline torch/csrc/autograd/functions/jit_closure.cpp]: I'm not sure how is that relevant in here

File: torch/csrc/autograd/functions/jit_closure.h [hunk 3]
```diff
@@ -13,6 +13,8 @@ namespace torch { namespace autograd {
 
 struct MultiStageClosure;
 
+// An AutogradClosureFactory caches the computation of a MultiStageClosure from
+// a Graph, so that creating an AutogradClosure is a ~no-op.
```
Comment (apaszke) [inline torch/csrc/autograd/functions/jit_closure.h]: I can't see how is that relevant for someone who's reading the header (and e.g. only cares about the interface and not about the internals of the factory).

File: torch/csrc/utils/pybind.h [hunk 4]
```diff
@@ -1,5 +1,6 @@
 #pragma once
 
+#include <Python.h>
```
Comment (apaszke) [inline torch/csrc/utils/pybind.h:3]: pybind already includes python IIRC

File: torch/csrc/autograd/functions/jit_closure.cpp [hunk 5]
```diff
@@ -41,8 +43,12 @@ struct Placeholder : public Function {
 // Used for inputs of previous previous stages
 struct PrevStageInput : public Replicate {};
 // Used for inputs to the closure and
+// (Execution root)
```
Comment (apaszke) [inline torch/csrc/autograd/functions/jit_closure.cpp]: ? Maybe inline this into the previous sentence?

File: torch/csrc/utils/pybind.h (hunk 4 above)
Comment (ezyang) [reply torch/csrc/utils/pybind.h:3]: This fixes a warning that occurs when Python.h is not the first include.

**Comment by ezyang** [comment]
Ready for you @apaszke

File: torch/csrc/autograd/functions/jit_closure.cpp [hunk 6]
```diff
@@ -40,9 +42,42 @@ struct Placeholder : public Function {
...
+// By what mechanism, then, can we actually pass the closure?  Here is the idea.
+// Instead of actually inserting an "Eval" node, we instead insert an
+// EvalPlaceholder, which doesn't know anything about evaluating a closure.
+// Then, at the time when we want to partially apply the actual closure
+// (computed from the forwards pass), we stick a pre hook on the EvalPlaceholder
+// that takes the inputs, does the actual Eval, and then passes on the outputs
+// (which the EvalPlaceholder subsequently passes through.)
+//
+// Remember that hooks are NOT stored on a Function object itself: they are
+// registered on a per graph basis.  So we can't do something like mutate a
```
Comment (apaszke) [inline torch/csrc/autograd/functions/jit_closure.cpp]: It's not per-graph. Each closure graph can have multiple AutogradClosures referencing it, and each has its own hooks

File: torch/csrc/autograd/functions/jit_closure.cpp [hunk 7]
```diff
@@ -40,9 +42,42 @@ struct Placeholder : public Function {
...
+// Eval Function to give it the autograd closure to run inside its main body:
+// that violates the invariant that autograd graphs are immutable!  (In other
+// words, the same EvalPlaceholder may be participating in multiple engine
+// executions) You truly must somehow associate these closures with the graph as
+// a whole, and there must be a mechanism to ferry this data to the Function
+// itself.  A hook is just the ticket.
```
Comment (apaszke) [inline torch/csrc/autograd/functions/jit_closure.cpp]: I'd prefer to call these callbacks in here. We have a distinction between hooks (registered on a Function) and callbacks (given as a Function -> std::function mapping to the engine)

File: torch/csrc/autograd/functions/jit_closure.cpp [hunk 8]
```diff
@@ -226,8 +261,55 @@ struct FusionGroupFunction : public Function {
...
+//    ]
+//    cur_stage_captures = [
+//      [%6.1, %3.1],
+//      [%23.1, %10.2],
+//      [] # Always empty!
+//    ]
 struct CrossStageStateDesc {
   CrossStageStateDesc(Graph* graph)
+    // FYI: graph->stage() == the number of stages we have traced
+    // (e.g., forwards+backwards = 2)
```
Comment (apaszke) [inline torch/csrc/autograd/functions/jit_closure.cpp]: That's not true. After forwards and backwards `graph->stage()` is still 1 (that's why I add 1 when resizing the vector)

File: torch/csrc/autograd/functions/jit_closure.cpp [hunk 9]
```diff
@@ -460,7 +559,7 @@ struct StageClosure {
         auto info = prev_stage_input_map.emplace(input_node, nullptr);
         auto & input_fn_ptr = info.first->second;
         // Create a node if insertion took place
-        if (info.second) input_fn_ptr.reset(new PrevStageInput());
+        if (info.second) input_fn_ptr.reset(new PrevStageInput()); // constructor is private!
```
Comment (apaszke) [inline torch/csrc/autograd/functions/jit_closure.cpp]: That's not true
